thinc/backends/numpy_ops.pyx

```diff
@@ -159,84 +156,143 @@ class NumpyOps(Ops):
         dX, d_params = backprop_lstm(dY, lengths, params, fwd_state)
         return dX, d_params
 
-    def maxout(self, const float[:, :, ::1] X):
-        cdef Pool mem = Pool()
+    def maxout(self, reals3d_ft X):
         cdef int B = X.shape[0]
         cdef int O = X.shape[1]
         cdef int P = X.shape[2]
 
-        cdef np.ndarray best = numpy.zeros((B, O), dtype='float32', order='C')
-        cdef np.ndarray which = numpy.zeros((B, O), dtype='int32', order='C')
-        if len(X) > 0:
-            cpu_maxout(<float*>best.data, <int*>which.data,
-                &X[0, 0, 0], B, O, P)
+        cdef np.ndarray best
+        cdef np.ndarray which = numpy.empty(shape=(B, O), dtype='int32', order='C')
+        if reals3d_ft is float3d_t:
+            best = numpy.empty(shape=(B, O), dtype="float32", order='C')
+            if len(X) > 0:
+                cpu_maxout(<float*>best.data, <int*>which.data,
+                    &X[0, 0, 0], B, O, P)
+        else:
+            best = numpy.empty(shape=(B, O), dtype="float64", order='C')
+            if len(X) > 0:
+                cpu_maxout(<double*>best.data, <int*>which.data,
+                    &X[0, 0, 0], B, O, P)
         return best, which
 
-    def backprop_maxout(self, const float[:, ::1] dY, int[:, ::1] which, int P):
+    def backprop_maxout(self, reals2d_ft dY, int[:, ::1] which, int P):
         cdef int B = dY.shape[0]
         cdef int O = dY.shape[1]
 
-        cdef np.ndarray dX = numpy.zeros((B, O, P), dtype='float32')
-        cpu_backprop_maxout(<float*>dX.data,
-            &dY[0, 0], &which[0, 0], B, O, P)
+        cdef np.ndarray dX
+        if reals2d_ft == float2d_t:
+            dX = numpy.zeros((B, O, P), dtype='float32')
+            cpu_backprop_maxout(<float*>dX.data, <float*>&dY[0, 0], &which[0, 0], B, O, P)
+        else:
+            dX = numpy.zeros((B, O, P), dtype='float64')
+            cpu_backprop_maxout(<double*>dX.data, <double*>&dY[0, 0], &which[0, 0], B, O, P)
+
         return dX
 
     def mish(self, np.ndarray X, threshold=20.0, inplace: bool = False):
         cdef np.ndarray Y
+
         if X.dtype == "float32":
-            if inplace:
-                Y = X
-            else:
-                Y = self.xp.empty_like(X)
-            cpu_mish(<float*>Y.data, <float *>X.data, threshold, X.size)
-            return Y
+            Y = _inplace_or_copy(X, inplace)
```

Comment:
The original code initialized `Y` to an empty array with same shape as `X` - but here we'd be copying `X` entirely, is that correct?

Reply:
Yes. The kernel now applies the activation in-place, which I did for all element-wise activations. It should not lead to any performance regressions, since most platforms vectorize `memcpy` and before this change we were doing scalar copies in the kernel itself (when values `>= threshold`).